src/codemirror/editor.ts

```diff
@@ -0,0 +1,301 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import * as CodeMirror
+  from 'codemirror';
+
+import 'codemirror/mode/meta';
+
+import {
+  ISignal, defineSignal
+} from 'phosphor/lib/core/signaling';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  AbstractCodeEditor, IPosition, IDimension, IEditorModel, IEditorConfiguration
+} from '../codeeditor/editor';
+
+import {
+  loadModeByFileName, loadModeByMIME
+} from '../codemirror';
+
+/**
+ * The class name added to CodeMirrorWidget instances.
+ */
+const EDITOR_CLASS = 'jp-CodeMirrorWidget';
+
+/**
+ * The name of the default CodeMirror theme
+ */
+export
+const DEFAULT_CODEMIRROR_THEME = 'jupyter';
+
+/**
+ * CodeMirror editor.
+ */
+export
+class CodeMirrorEditor extends AbstractCodeEditor implements IEditorModel, IEditorConfiguration {
```

Comment:
The design implied by this line is a bit awkward as `AbstractCodeEditor` doesn't implement any of the interfaces listed here. That means that the `CodeMirrorEditor` is implementing 2 interfaces and more-or-less another one through `AbstractCodeEditor`. In general we try to keep concrete classes to implementing a single interface, unless the other ones follow a mixin pattern.

Some ideas:

* Maybe `AbstractCodeEditor` could implement `IEditorModel`?
* Maybe `AbstractCodeEditor` could have a configuration property that has a type of `EditorConfiguration`. In other words, use containment for that part, rather than implements.